### PR TITLE
Return Server produced error if Http status code is 403

### PR DESCRIPTION
**Pull request by qdpham13**
https://buganizer.corp.google.com/issues/268398841
Pass back server error if it is the 403 API enablement response.

### CODE REVIEW

File: firebase-config/src/main/java/com/google/firebase/remoteconfig/internal/ConfigRealtimeHttpClient.java [hunk 1]
```diff
@@ -470,12 +489,17 @@ public void beginRealtimeHttpStream() {
       if (connectionFailed || responseCode == HttpURLConnection.HTTP_OK) {
         retryHttpConnectionWhenBackoffEnds();
       } else {
+        String errorMessage =
+            String.format(
+                "Unable to connect to the server. Try again in a few minutes. Http Status code: %d",
+                responseCode);
+        if (responseCode == 403) {
```
Comment (danasilver) [inline firebase-config/src/main/java/com/google/firebase/remoteconfig/internal/ConfigRealtimeHttpClient.java:496]: What if we get a 403 without a message we expect? We should still return some error message in the exception.

File: firebase-config/src/main/java/com/google/firebase/remoteconfig/internal/ConfigRealtimeHttpClient.java [hunk 2]
```diff
@@ -470,12 +489,17 @@ public void beginRealtimeHttpStream() {
       if (connectionFailed || responseCode == HttpURLConnection.HTTP_OK) {
         retryHttpConnectionWhenBackoffEnds();
       } else {
+        String errorMessage =
+            String.format(
+                "Unable to connect to the server. Try again in a few minutes. Http Status code: %d",
```
Comment (danasilver) [inline firebase-config/src/main/java/com/google/firebase/remoteconfig/internal/ConfigRealtimeHttpClient.java:494]: Nit: NTTP is capitalized

**Comment by danasilver** [comment]
Any tests we can add for this?

File: firebase-config/src/test/java/com/google/firebase/remoteconfig/FirebaseRemoteConfigTest.java [hunk 3]
```diff
@@ -1332,6 +1332,23 @@ public void realtime_successfulStatusCode_doesNotIncreaseConfigMetadataBackoffDa
...
+        .when(configRealtimeHttpClientSpy)
+        .closeRealtimeHttpStream(any(HttpURLConnection.class));
+    when(mockHttpURLConnection.getErrorStream())
+        .thenReturn(new ByteArrayInputStream("{ \"code\": 403 }".getBytes(StandardCharsets.UTF_8)));
+    when(mockHttpURLConnection.getResponseCode()).thenReturn(403);
+    configRealtimeHttpClientSpy.beginRealtimeHttpStream();
+
+    verify(configRealtimeHttpClientSpy, never()).startAutoFetch(any());
+    verify(configRealtimeHttpClientSpy, never()).retryHttpConnectionWhenBackoffEnds();
+    verify(mockStreamErrorEventListener).onError(any(FirebaseRemoteConfigServerException.class));
```
Comment (danasilver) [inline firebase-config/src/test/java/com/google/firebase/remoteconfig/FirebaseRemoteConfigTest.java:1349]: Maybe add the message we expect from the backend and verify it's parsed correctly?

File: firebase-config/src/main/java/com/google/firebase/remoteconfig/internal/ConfigRealtimeHttpClient.java [hunk 4]
```diff
@@ -423,6 +423,7 @@ private String parseErrorResponseMessage(InputStream inputStream) {
       }
     } catch (IOException ex) {
       // Unable to parse error message.
+      return "Unable to connect to the server, access is forbidden. Http Status code: 403";
```
Comment (danasilver) [inline firebase-config/src/main/java/com/google/firebase/remoteconfig/internal/ConfigRealtimeHttpClient.java:426]: Same "HTTP" capitalization here. Would also call this method `parseForbiddenResponseMessage` if it's 403-specific and add a comment why we're trying to parse it for future reference

File: firebase-config/src/main/java/com/google/firebase/remoteconfig/internal/ConfigRealtimeHttpClient.java [hunk 5]
```diff
@@ -412,18 +412,21 @@ private boolean isStatusCodeRetryable(int statusCode) {
...
-  private String parseErrorResponseMessage(InputStream inputStream) {
+  private String parseForbiddenErrorResponseMessage(InputStream inputStream) {
     StringBuilder response = new StringBuilder();
 
     try {
       BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(inputStream));
       String message = "";
       while ((message = bufferedReader.readLine()) != null) {
         response.append(message);
+        Log.i(TAG, message);
```
Comment (danasilver) [inline firebase-config/src/main/java/com/google/firebase/remoteconfig/internal/ConfigRealtimeHttpClient.java:423]: Extra log?

**Review by danasilver — APPROVED** [review]

File: firebase-config/src/main/java/com/google/firebase/remoteconfig/internal/ConfigRealtimeHttpClient.java (hunk 5 above)
Comment (qdpham13) [reply firebase-config/src/main/java/com/google/firebase/remoteconfig/internal/ConfigRealtimeHttpClient.java]: Good catch, thanks!

**Pull request merged**